### ISSUE TITLE
A few questions about akka-ddd status

### ISSUE THREAD

**Issue by onsails** [pawelkaczor/akka-ddd#9]
Hello,
We in [Actor](https://github.com/actorapp/actor-platform) are considering implementing a database-agnostic persistence layer. We are going to use event-sourcing for everything, but it leads us to a question: how we should build dashboards, do analytics, if we don't have a consistent data representation available for use by third-party systems?
We were considering implementing our own solution for synchronizing state of Persistent Actors (like User and Groups names, Group participants, User balances, etc.). But today I've stumbled upon akka-ddd and found out that you already have solutions for this, and now I'm considering an ability to modify akka-ddd for our needs and use it in actor-platform. And contribute our modifications to akka-ddd project too of course.
But before we finally decide use akka-ddd there are some questions:

– Some of our deployments use Cassandra for events persistence. Do I understand right that `akka-persistence-*` modules needs to be extended for compatibility with akka-ddd? Is there a way to completely replace `eventstore-akka-persistence` by another `akka-persistence-*` module?
– We are using protobuf in our client-server protocol, we are migrating our internal server's serialization to it too, using [ScalaPB](https://trueaccord.github.io/ScalaPB). It is backward- and forward-compatible, fast and produces small-sized serialized data. Is there a way to use custom serialization with akka-ddd?

Thanks for your time, looking forward hearing your feedback!

**Comment by pawelkaczor** [comment]
Hello Andrey,

> > Is there a way to completely replace eventstore-akka-persistence by another akka-persistence-\* module?

Currently [Akka Persistence](http://doc.akka.io/docs/akka/snapshot/scala/persistence.html) does not support queries (like: give me all events for particular Aggregate Root class). Akka-DDD resolves this problem by using EventStore's projections. The consequence are:
- `view-update` module depends on `eventstore-akka-persistence` module (**so you can not just replace `eventstore-akka-persistence` module with another one**).
- if you want to use Saga or Receptor pattern you must provide implementation of EventStreamSubscriber (contained in `akka-ddd-messaging` module) interface. Currently the only implementation of EventStreamSubscriber exists in `eventstore-akka-persistence` module.

**Suggested approach**
Quite recently Akka team has started to work on [Improve the Query Side](https://github.com/akka/akka/issues/16541) issue. Looks like they have decided to ship first version of this functionality with Akka 2.4-M3 (next milestone of Akka 2.4!). Therefore I would suggest you to consider Akka 2.4 as your target platform if your concern is database agnosticism of the `view-update` functionality. I believe Akka-DDD could incorporate query-side goodies coming with Akka 2.4.

> > Is there a way to use custom serialization with akka-ddd?

Core functionality of Akka-DDD is not tied to particular serialization format except functionality provided in pl.newicom.dddd.process package (Receptor and Saga patterns). If you take look at pl.newicom.dddd.office.OfficeInfo you will find JsonSerializationHints class being used. OfficeInfo is used by Receptor/Saga and `view-update`.

I wish you good luck with your project and hope you can benefit from Akka-DDD and/or Akka-Persistence now or in the near future.

**Comment by onsails** [comment]
So I think we will implement things manually, looking back to akka-ddd. Thank you for such informative explaination!

**Comment by stuartsaltzman** [comment]
@prettynatty - have you made any progress in incorporating Cassandra in akka-ddd? Thanks!

**Comment by onsails** [comment]
We ended up with our own solution based on Akka Persistence. 
Actually we separated C and Q by dumbly putting them into traits, both are mixed into one actor.  We had to do this because we encountered bugs in PersistentViews. We are going to do a real separation after moving to Akka 2.4.0 – it has great opportunities like persistent streams.
Also we've discovered [another one](https://github.com/NextGenTel/akka-tools/tree/master/akka-tools-persistence) project. We liked their simple abstractions and may be we will port them to our CQRS-ish solution.
